### PR TITLE
Added merge strategy to DEVELOPMENT.md

### PR DESCRIPTION
**Pull request by youngchen7**
@kokosing

### CODE REVIEW

File: DEVELOPMENT.md [hunk 1]
```diff
@@ -87,6 +87,16 @@ Maintain the same quality for production and test code.
 
 When writing a Git commit message, follow these [guidelines](https://chris.beams.io/posts/git-commit/).
 
+### Git merge strategy
+
+We prefer to `rebase and merge`, so cleaning up and squashing commits on the feature branch is best.
+`squash and merge` is reserved for when git is unfamiliar to the external contributor.
```
Comment (kokosing) [inline DEVELOPMENT.md:93]: `for when` does not sound like a proper english. I think we could skip the part of `squash and merge`, it is exceptional case, we should not learn people how to leverage this path.

File: DEVELOPMENT.md [hunk 2]
```diff
@@ -87,6 +87,16 @@ Maintain the same quality for production and test code.
...
 
+### Git merge strategy
+
+We prefer to `rebase and merge`, so cleaning up and squashing commits on the feature branch is best.
+`squash and merge` is reserved for when git is unfamiliar to the external contributor. 
+We care a lot about the history, so this is why how change is structured between commits and what the
+commit messages are is also a part of code review. However we prefer not to merge anything where changes goes
+back and forth between commits in the same PR. This makes the entire git history much cleaner. Notice that
+writing code is creative process and it is natural to got back and forth when writing the code, however
+when visiting history such small steps are not helpful.
```
Comment (kokosing) [inline DEVELOPMENT.md:98]: In general content is ok to me, but it requires some language improvements from native speakers.

@mosabua or @electrum Would you like to help here?

File: DEVELOPMENT.md [hunk 3]
```diff
@@ -87,6 +87,16 @@ Maintain the same quality for production and test code.
...
 When writing a Git commit message, follow these [guidelines](https://chris.beams.io/posts/git-commit/).
 
+### Git merge strategy
+
+We prefer to `rebase and merge`, so cleaning up and squashing commits on the feature branch is best.
+`squash and merge` is reserved for when git is unfamiliar to the external contributor. 
+We care a lot about the history, so this is why how change is structured between commits and what the
+commit messages are is also a part of code review. However we prefer not to merge anything where changes goes
+back and forth between commits in the same PR. This makes the entire git history much cleaner. Notice that
+writing code is creative process and it is natural to got back and forth when writing the code, however
```
Comment (findinpath) [inline DEVELOPMENT.md:97]: `to got back` -> `to go back`

**Review by kokosing — COMMENTED** [review]
@trinodb/maintainers @ksobolew FYI

File: DEVELOPMENT.md [hunk 4]
```diff
@@ -87,6 +87,16 @@ Maintain the same quality for production and test code.
 
 When writing a Git commit message, follow these [guidelines](https://chris.beams.io/posts/git-commit/).
 
+### Git merge strategy
+
+We prefer to `rebase and merge`, so cleaning up and squashing commits on the feature branch is best.
```
[the remark this answers is not in the document]
Comment (hashhar) [reply DEVELOPMENT.md:92]: While we are at it would it be helpful to point to something like https://fle.github.io/git-tip-keep-your-branch-clean-with-fixup-and-autosquash.html too so that the next paragraph makes sense. (It talks about keeping commits structured and changes not going back and forth within commits).

[the remark this answers is not in the document]
Comment (youngchen7) [reply DEVELOPMENT.md:92]: What link would you prefer? https://docs.github.com/en/github/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges This one seems like a pretty good starting point.

@hashhar I think the doc you linked isn't so much an explanation of squash & merging as it is a tip on specific git tools (fixup and autosquash). For me, I typically use git commit --amend when fixing up the previous commit. Maybe a more generic article like this: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History would be better? Let me know your thoughts!

[the remark this answers is not in the document]
Comment (hashhar) [reply DEVELOPMENT.md:92]: I mentioned fixup and autosquash mainly because in actual practice most of the times people run into a state where the commit they want to edit isn't the previous commit - it's something older.

e.g. 3 commits A, B and C. Reviewers asks to change something in A. You add that as a fixup commit to A so that in the end you can autosquash and git will handle which commits go where during the rebase.

However - now thinking about it I feel we can leave this detail out - it's more likely to confuse new people and if anyone or any PR needs this then the reviewers can point people in the direction.

File: DEVELOPMENT.md (hunk 1 above)
Comment (hashhar) [reply DEVELOPMENT.md]: Squash and merge is enabled but it does mess up the release notes checklist for some reason (because the comitter is not the person who merged the PR?). It's there for "exceptional" cases.

IMO best to remove mention of `squash and merge` as Grzegorz said.

File: DEVELOPMENT.md [hunk 5]
```diff
@@ -87,6 +87,15 @@ Maintain the same quality for production and test code.
...
+We care a lot about the history, so this is why how change is structured between commits and what the
+commit messages are is also a part of code review. However we prefer not to merge anything where changes go
+back and forth between commits in the same PR. This makes the entire git history much cleaner. Notice that
+writing code is a creative process and it is natural to go back and forth when writing code, however
+when visiting history such small steps are not helpful.
```
Comment (alexjo2144) [inline DEVELOPMENT.md:97]: Here's my stab at a rewording:

```
### Git merge strategy
PRs are usually merged into `master` using the  `rebase and merge` strategy. 

It is important to keep commits on feature branches neat, squashing the feature branch as necessary.
Commit messages and  history are also important, as they are used by other developers to keep track of the motivation behind changes.
Keep logical diffs grouped together in separate commits, and order commits in a way that explains
the progress of the changes. Rewriting and reordering commits may be a necessary part of the PR review
process, as the code changes.
```

Comment (mosabua) [reply DEVELOPMENT.md:97]: This better .. I would go even further and mention that typical PRs should strive to just contain one commit.. 

I am not aware of an actual policy beyond that .. since "neat" and "logical" and such are all very subjective terms and are not that actionable.

Comment (findepi) [reply DEVELOPMENT.md:97]: > typical PRs should strive to just contain one commit..


that can cause people to create too big commits, let's not write that

Comment (kasiafi) [reply DEVELOPMENT.md:97]: > typical PRs should strive to just contain one commit..

I don't think that's the case. Contrarily, we often ask contributors to extract unrelated changes (e.g. fix-typos, reformats) to separate commits.
The rule is that a commit should be a logically separate chunk of work. 

One thing that could be probably mentioned is that if a PR consists of multiple commits, it is expected that every prefix of it is  _correct_. That is, there might be preparatory commits at the bottom of the stack that don't bring any value by themselves, but none of the commits should introduce an error that is fixed by some future commit.

Comment (ksobolew) [reply DEVELOPMENT.md:97]: > it is expected that every prefix of it is correct

I like that rule 👍

**Comment by kokosing** [comment]
@youngchen7 We captured a lot of good insights from the community, can you please update?

**Comment by youngchen7** [comment]
Thanks for all the feedback! I've updated the description to incorporate @alexjo2144's rewording, and added the discussion points to the end.

**Comment by youngchen7** [comment]
Rebased and squashed per the additions to DEVELOPMENT.md in this PR 😄

File: DEVELOPMENT.md [hunk 6]
```diff
@@ -87,6 +87,23 @@ Maintain the same quality for production and test code.
 
 When writing a Git commit message, follow these [guidelines](https://chris.beams.io/posts/git-commit/).
 
+### Git merge strategy
+
+PRs are usually merged into `master` using the  `rebase and merge` strategy.
+
+It is important to keep commits on feature branches neat, squashing the feature branch as necessary.
```
Comment (mosabua) [inline DEVELOPMENT.md:94]: I would either remove this paragraph or move it to the end of this section..

Comment (youngchen7) [reply DEVELOPMENT.md]: Moved the paragraph to the end - I think the context it provides is good to have.

File: DEVELOPMENT.md [hunk 7]
```diff
@@ -87,6 +87,23 @@ Maintain the same quality for production and test code.
 
 When writing a Git commit message, follow these [guidelines](https://chris.beams.io/posts/git-commit/).
 
+### Git merge strategy
+
+PRs are usually merged into `master` using the  `rebase and merge` strategy.
```
Comment (kokosing) [inline DEVELOPMENT.md:92]: `PR` -> `Pull request`. It is more or less official documentation so let's be crystal clear.

Comment (kokosing) [inline DEVELOPMENT.md:92]: Please add a link to Github that explains what `Rebase and merge` is. https://docs.github.com/en/github/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#rebase-and-merge-your-pull-request-commits

File: DEVELOPMENT.md [hunk 8]
```diff
@@ -87,6 +87,23 @@ Maintain the same quality for production and test code.
...
+
+A typical PR should strive to just contain one commit - logically unrelated changes should generally be extracted
+into their own PRs.
+
+If a PR does consist of multiple commits, it is expected that every prefix of it is correct. That is,
+there might be preparatory commits at the bottom of the stack that don't bring any value by themselves,
+but none of the commits should introduce an error that is fixed by some future commit.
+
+It is important to keep commits on feature branches neat, squashing the feature branch as necessary.
+Commit messages and  history are also important, as they are used by other developers to keep track of
```
Comment (kokosing) [inline DEVELOPMENT.md:102]: double space

File: DEVELOPMENT.md [hunk 9]
```diff
@@ -87,6 +87,23 @@ Maintain the same quality for production and test code.
...
+the motivation behind changes. Keep logical diffs grouped together in separate commits, and order commits
+in a way that explains the progress of the changes. Rewriting and reordering commits may be a
+necessary part of the PR review process as the code changes.
+
```
Comment (kokosing) [inline DEVELOPMENT.md:106]: Please also mention that mechanical changes (refactorings) like class,method,variable renames, moving code around should be always separated from logical changes. These two should be stored in separate commits, they should never be mixed. As this makes code review much harder and easier to introduce unintended behavior change (possibly bugs).

// I am not sure how to write it nicely in english.

Comment (youngchen7) [reply DEVELOPMENT.md:110]: Reworded this a bit and added it at the end. Let me know how this looks to you. 
> class,method,variable renames, moving code around
I think all of these falls under the category of "refactoring", so I just specified that. Unless there are mechanical changes you can think of that don't qualify as refactoring, I may consider describing it as refactor vs logical changes.

File: DEVELOPMENT.md [hunk 10]
```diff
@@ -87,6 +87,27 @@ Maintain the same quality for production and test code.
...
+### Git merge strategy
+
+Pull requests are usually merged into `master` using the  [`rebase and merge`](https://docs.github.com/en/github/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#rebase-and-merge-your-pull-request-commits) strategy.
+
+A typical pull request should strive to just contain one commit - logically unrelated changes should generally be extracted
+into their own PRs.
+
+If a pull request does consist of multiple commits, it is expected that every prefix of it is correct. That is,
+there might be preparatory commits at the bottom of the stack that don't bring any value by themselves,
+but none of the commits should introduce an error that is fixed by some future commit.
```
Comment (hashhar) [inline DEVELOPMENT.md:99]:
```suggestion
each of the commits should build successfully and pass the tests. For this reason the commit adding a feature or fixing a bug should also contain the tests instead of them being separate commits.
```

I feel this example makes it more obvious what the text is trying to say in my opinion.

**Review by hashhar — APPROVED** [review]
LGTM from me % comment.

File: DEVELOPMENT.md (hunk 10 above)
Comment (youngchen7) [reply DEVELOPMENT.md:99]: I think these two descriptions are similar but not equivalent. What if a feature is added, and then a later commit adds or improves the unit test suite? Are we enforcing that these should be the same commit? Adding or improving unit tests later seems fine from my POV but let me know what you think.

Comment (hashhar) [reply DEVELOPMENT.md:99]: Good point. The existing text is more precise and doesn't leave room for confusion. Let's keep as is.

File: DEVELOPMENT.md [hunk 11]
```diff
@@ -87,6 +87,27 @@ Maintain the same quality for production and test code.
...
+
+It is important to keep commits on feature branches neat, squashing the feature branch as necessary.
+
+Commit messages and  history are also important, as they are used by other developers to keep track of
```
Comment (kokosing) [inline DEVELOPMENT.md:103]: double space

**Review by kokosing — APPROVED** [review]
Thank you!

**Comment by kokosing** [comment]
@trinodb/maintainers Any other concerns? Otherwise I would like to merge it.

File: DEVELOPMENT.md (hunk 11 above)
Comment (youngchen7) [reply DEVELOPMENT.md]: Fixed - also, checked the file for double spaces and there are a lot of double spaces after periods. Is that the correct convention?

Comment (kokosing) [reply DEVELOPMENT.md]: We should remove these. Can you please fix them in separate commit?

**Comment by kokosing** [comment]
@youngchen7 and others, thank you for working on this!
